tests/testsuite/build_dir.rs

```diff
@@ -0,0 +1,193 @@
+        .run();
+
+    let docs_dir = p.root().join("build/doc");
+    assert!(docs_dir.exists());
+    assert!(docs_dir.join("foo/index.html").exists());
+}
+
+#[cargo_test]
+fn cargo_package_should_output_to_target_dir() {
+    let p = project()
+        .file("src/main.rs", r#"fn main() { println!("Hello, World!") }"#)
+        .file(
+            ".cargo/config.toml",
+            r#"
+            [build]
+            target-dir = "build"
+            "#,
+        )
+        .build();
+
+    p.cargo("package")
+        .masquerade_as_nightly_cargo(&[])
+        .enable_mac_dsym()
+        .run();
+
+    assert_build_dir(p.root().join("build"), "debug", true);
+
+    let package_dir = p.root().join("build/package");
+    assert!(package_dir.exists());
+    assert!(package_dir.join("foo-0.0.1.crate").exists());
+    assert!(package_dir.join("foo-0.0.1.crate").is_file());
+    assert!(package_dir.join("foo-0.0.1").exists());
+    assert!(package_dir.join("foo-0.0.1").is_dir());
+}
+
+#[cargo_test]
+fn cargo_clean_should_clean_the_target_dir() {
+    let p = project()
+        .file("src/main.rs", r#"fn main() { println!("Hello, World!") }"#)
+        .file(
+            ".cargo/config.toml",
+            r#"
+            [build]
+            target-dir = "build"
+            "#,
+        )
+        .build();
+
+    p.cargo("build")
+        .masquerade_as_nightly_cargo(&[])
+        .enable_mac_dsym()
+        .run();
+
+    assert_build_dir(p.root().join("build"), "debug", true);
+
+    p.cargo("clean")
+        .masquerade_as_nightly_cargo(&[])
+        .enable_mac_dsym()
+        .run();
+
+    assert!(!p.root().join("build").exists());
+}
+
+#[track_caller]
+fn assert_build_dir(path: PathBuf, profile: &str, is_build_dir: bool) {
+    println!("checking if {path:?} is a build directory ({is_build_dir})");
+    // For things that are in both `target` and the build directory we only check if they are
+    // present if `is_build_dir` is true.
+    if is_build_dir {
+        assert_eq!(is_build_dir, path.join(profile).is_dir());
+    }
+
+    let error_message = |dir: &str| {
+        if is_build_dir {
+            format!("`{dir}` dir was expected but not found")
+        } else {
+            format!("`{dir}` dir was not expected but was found")
+        }
+    };
+
+    assert_eq!(
+        is_build_dir,
+        path.join(profile).join("deps").is_dir(),
+        "{}",
+        error_message("deps")
+    );
+    assert_eq!(
+        is_build_dir,
+        path.join(profile).join("build").is_dir(),
+        "{}",
+        error_message("build")
+    );
+    assert_eq!(
+        is_build_dir,
+        path.join(profile).join("incremental").is_dir(),
+        "{}",
+        error_message("incremental")
+    );
+}
+
+#[track_caller]
+fn assert_executeable_exists(path: PathBuf, name: &str, exists: bool) {
+    let executable = if cfg!(target_os = "windows") {
+        format!("{name}.exe")
+    } else {
+        name.to_string()
+    };
```

Comment:
You can always insert https://doc.rust-lang.org/std/env/consts/constant.EXE_SUFFIX.html, making this not have to deal with conditional compilation